q2cli/builtin/tools.py

```diff
@@ -80,40 +80,149 @@ def export_data(input_path, output_path, output_format):
+            click.echo()
+
+
+def deformat_description(description):
+    import re
+    deformatted = re.sub(r"[\t\n]+", ' ', description)
+    despaced = re.sub(r" +", ' ', deformatted)
+    return despaced
+
+
+def get_matches(words, possibilities, cutoff=0.5):
+    from difflib import get_close_matches
+    matches = []
+    for word in words:
+        matches += get_close_matches(word,
+                                     possibilities,
+                                     n=len(possibilities),
+                                     cutoff=cutoff)
+        # substring search
+        if cutoff != 1:
+            for possibility in possibilities:
+                if word.lower() in possibility.lower():
+                    if possibility not in matches:
+                        matches.append(possibility)
+
+    return matches
+
+
+@tools.command(
+        name='show-types',
+        help='List the available semantic types.',
+        short_help='',
+        cls=ToolCommand
+)
+@click.argument('types', nargs=-1)
+@click.option('--strict', is_flag=True,
+              help='Show only exact matches for the type argument(s).')
+@click.option('--tsv', is_flag=True,
+              help='Print as machine readable tab separated values.')
+def show_types(types, strict, tsv):
+    pm = q2cli.util.get_plugin_manager()
+
+    if types and strict:
+        matches = get_matches(types, list(pm.artifact_classes), 1)
+    elif types:
+        matches = get_matches(types, list(pm.artifact_classes))
+    else:
+        matches = list(pm.artifact_classes)
+
+    descriptions = {}
+    for match in sorted(matches):
+        description = pm.artifact_classes[match].description
+        descriptions[match] = description
+
+    print_descriptions(descriptions, tsv)
+
+
+@tools.command(
+        name='show-formats',
+        help='List the availabe formats.',
+        short_help='',
+        cls=ToolCommand
+)
+@click.argument('formats', nargs=-1)
+@click.option('--importable', is_flag=True,
+              help='List the importable formats.')
+@click.option('--exportable', is_flag=True,
+              help='List the exportable formats.')
+@click.option('--strict', is_flag=True,
+              help='Show only exact matches for the format argument(s).')
+@click.option('--tsv', is_flag=True,
+              help='Print as machine readable tab separated values.')
+def show_formats(formats, importable, exportable, strict, tsv):
+    if importable and exportable:
+        raise click.UsageError("'--importable' and '--exportable' flags are "
+                               "mutually exclusive.")
+    if not importable and not exportable:
+        raise click.UsageError("One of '--importable' or '--exportable' flags "
+                               "is required.")
+
+    pm = q2cli.util.get_plugin_manager()
+    portable_formats = pm.importable_formats if importable \
+        else pm.exportable_formats
+    available_formats = list(portable_formats)
```

Comment:
It looks like you're defining a second variable here so that you have this information in a list (`available_formats`) and a dict (`portable_formats`). It's a little confusing to read though, as it's not immediately clear why you need need both of these. I'm guessing it's to provide a list to `get_matches`. If that's correct, calling:

```
 get_matches(formats, portable_formats.keys(), ...)
```

would be a little clearer in my opinion.